### DISCUSSION TITLE
Writing a library crate that includes both rust and slint?

### DISCUSSION THREAD

**Discussion by andrew-otiv** [slint-ui/slint#4623] [Q&A]
Is it possible to make a rust library crate that contains both slint and rust code, and is intended to be re-usable in multiple projects? Suppose you had some custom control for slint that you want to push to crates.io... i.e. a video streaming plugin or a slippy map plugin... is that currently feasible?

**Comment by andrew-otiv** [comment]
We're working on refactoring the gstreamer example in this direction; I pushed it here and made a Draft PR: https://github.com/slint-ui/slint/pull/4627

We haven't managed to resolve ownership and thread safety issues yet. We had to jump through some hoops with closures since most of Slint's types are code generated under names that would not be known to a library (and they don't implement trais exposing what we need either).

`NonNull<i_slint_core::sharedvector::SharedVectorInner<rgb::RGB<u8>>>` cannot be shared between threads safely
`Rc<RefCell<usvg_tree::ClipPath>>` cannot be sent between threads safely
`Rc<tiny_skia_path::path::Path>` cannot be sent between threads safely
`NonNull<vtable::vrc::VRcInner<'static, i_slint_core::item_tree::ItemTree_vtable_mod::ItemTreeVTable, InnerApp>>` cannot be shared between threads safely

Does this seem like a promising approach? Maybe we should write this as a macro that spits out the code we already had working, despite not knowing what types it will actually be using. Crates can ship macros, and slint is macro based anyway. We're just new to writing rust macros.

**Reply by ogoffart** [reply]
Unfortunately, as of now, the Image cannot be sent between threads, because it uses caches.
But the SharedPixelBuffer is Send

**Comment by ogoffart** [comment]
Having library in rust is something we'd like to allow, but unfortunately this is not really supported right now.
The main problem is that the types are generated by the Slint build script or macro and are all locale to the end crate. We would need to somehow tell the Slint compiler that some types from some of the files are going to be in a different crate.
We still don't have clean solutions for that.

(Some workaround could be done with rust macro or such, but we need a better way.)

**Reply by andrew-otiv** [reply]
Is it on your roadmap, or is it not a priority for Slint's niche? Just to help decide if we implement workarounds or wait. For context, we want to write a couple smaller libraries and share them internally between projects within our company.

**Reply by tronical** [reply]
This is on our roadmap, the ability to encapsulate a bunch of controls written with .slint files and associated Rust or C++ code is in scope. #2031 is one particular item towards it. In the meantime, one thing that you can do is to use the library feature like coop widgets does in https://codeberg.org/flovansl/co_sl/src/branch/main/widgets/coop/  - for your .slint files, and organise native code around singletons that you could initialise with helper functions: https://github.com/slint-ui/slint/blob/b4e36f1174e394ca8a8c5bda93a1f1852f300693/examples/virtual_keyboard/rust/main.rs#L14

Not ideal but probably the best we've got for now :(

**Comment by sinkyl** [comment]
Hi!
Any update on this?

-- nonsense removed --

I started a new project in Rust within a workspace where I have 3 crates, one a library and the other two are bin projects. I thought I could create all my widgets in the library and use the one I needed in the bin projects.
